posts/inside-rust/2022-08-11-diagnostic-effort.md

```diff
@@ -0,0 +1,336 @@
+`Span` for us:
+
+```rust=
+struct ReturnTypeArrow {
+    span: Span,
+}
+```
+
+In most cases, this will just be the `Span`s that are used by the original
+diagnostic emission logic and values that are interpolated into diagnostic
+messages.
+
+After that, we should add the derive, add our error attribute and annotate the
+primary `Span` (that was given to `struct_span_err`).
+
+```rust=
+#[derive(SessionDiagnostic)]
+#[error(parser::return_type_arrow)]
+struct ReturnTypeArrow {
+    #[primary_span]
+    span: Span,
+}
+```
+
+Each diagnostic should have a unique slug. By convention, these always start
+with the crate that the error is related to (`parser` in this example). This
+slug will be used to find the actual diagnostic message in our translation
+resources, which we'll see shortly.
+
+Finally, we need to add any labels, notes, helps or suggestions:
+
+```rust=
+#[derive(SessionDiagnostic)]
+#[error(parser::return_type_arrow)]
+struct ReturnTypeArrow {
+    #[primary_span]
+    #[suggestion(applicability = "machine-applicable", code = "->")]
+    span: Span,
+}
+```
+
+In this example, there's just a single suggestion - to replace the `:` with
+a`->`.
+
+Before we're finished, we have to [add the diagnostic messages to the
+translation resources..](#Adding-translation-resources…)
+
+For more documentation on diagnostic derives, see the [diagnostic structs
+chapter of the `rustc` dev guide][diag_struct].
+
+#### Manually implementing `SessionDiagnostic`...
+Some diagnostics are too complicated to be generated from a diagnostic type
+using the diagnostic derive. In these cases, `SessionDiagnostic` can be
+implemented manually.
+
+Using the same type as in ["Using a diagnostic
+derive..."](#Using-a-diagnostic-derive…), we can implement `SessionDiagnostic`
+as below:
+
+```rust=
+use rustc_errors::{fluent, SessionDiagnostic};
+
+struct ReturnTypeArrow { span: Span }
+
+impl SessionDiagnostic for ReturnTypeArrow {
+    fn into_diagnostic(self, sess: &'_ rustc_session::Session) -> DiagnosticBuilder<'_> {
+        sess.struct_span_err(
+            self.span,
+            fluent::parser::return_type_arrow,
+        )
+        .span_suggestion_short(
+            self.span,
+            fluent::parser::suggestion,
+            "->".to_string(),
+            Applicability::MachineApplicable,
+        )
+    }
+}
+```
+
+Instead of using strings for the messages as in the original diagnostic
+emission logic, typed identifiers referring to translation resources are used.
+Now we just have to [add the diagnostic messages to the translation
+resources..](#Adding-translation-resources…).
+
+#### Examples
+For more examples of diagnostics ported to use the diagnostic derive or written
+manually, see the following pull requests:
+
+- https://github.com/rust-lang/rust/pull/98353
+- https://github.com/rust-lang/rust/pull/98415
+- https://github.com/rust-lang/rust/pull/97093
+- https://github.com/rust-lang/rust/pull/99213
```

Comment:
I was going to ask if there were some example PRs people could look at. Thanks for including these!